feat: Add ExtProc support to BackendTrafficPolicy for backend-specific external processing

shiavm006 opened the pull request:
Summary
Implements backend-specific external processing by adding ExtProc support to BackendTrafficPolicy. This enables different external processing logic per backend after backend selection occurs.

>Changes
- Add `extProc` field to `BackendTrafficPolicySpec` 
- Implement `buildBackendExtProc` function that reuses existing ExtProc translation logic
- Add ExtProc support to IR TrafficFeatures for backend-specific processing
- Update generated code and documentation

Fixes: Support Ext Proc that executes per Backend 
issue fixed #5881
what r ur views on it ?

guydc commented:
Hi @shiavm006. Thanks for looking into this feature request. 
In the linked issue, the proposal for API was: 
> Ext Proc https://gateway.envoyproxy.io/docs/api/extension_types/#extproc can be enhanced with a phase/stage field so it can be executed in the Route / Backend or All phases/stages

In other words, the proposal was to enhance the existing EnvoyExtensionPolicy and make it possible to configure the scope of the ext-proc filter: downstream (listener HCM filter), upstream (cluster filter) or both. 

I support that API proposal:
- It keeps control over extensions in the hands of a privileged user. BackendTrafficPolicy can be used (typically) by application developers (owners of XRoutes), who are not as trusted when it comes to injecting extensions. 
- It's consistent with the current API: 
  - The current ext-proc is not in ClientTrafficPolicy
  - Ext-proc is the same CRD as all other extensions (lua, wasm, ... )

shiavm006 commented:
> Hi @shiavm006. Thanks for looking into this feature request. In the linked issue, the proposal for API was:
> 
> > Ext Proc https://gateway.envoyproxy.io/docs/api/extension_types/#extproc can be enhanced with a phase/stage field so it can be executed in the Route / Backend or All phases/stages
> 
> In other words, the proposal was to enhance the existing EnvoyExtensionPolicy and make it possible to configure the scope of the ext-proc filter: downstream (listener HCM filter), upstream (cluster filter) or both.
> 
> I support that API proposal:
> 
> * It keeps control over extensions in the hands of a privileged user. BackendTrafficPolicy can be used (typically) by application developers (owners of XRoutes), who are not as trusted when it comes to injecting extensions.
> * It's consistent with the current API:
>   
>   * The current ext-proc is not in ClientTrafficPolicy
>   * Ext-proc is the same CRD as all other extensions (lua, wasm, ... )

hi @guydc Thank you for the clarification! You're absolutely right about the security model concerns.
I see now that my BackendTrafficPolicy approach has architectural issues
but i have some question i would apprecaite if u clarify that 
1. Field name preference: phase, stage, or executionPoint?
2. Default value for backwards compatibility (I assume "downstream")?
Thanks for your patience and the clear guidance!

arkodg commented:
@shiavm006 i recommended raising a API only PR for EEP, you could use `stage` , the orig issue has possible enum options, `Route` is the default

shiavm006 commented:
> @shiavm006 i recommended raising a API only PR for EEP, you could use `stage` , the orig issue has possible enum options, `Route` is the default

Hi  ,
Thank you for the detailed feedback and clear guidance! I've updated the PR to follow the approach you recommended.
I have removed the previous implementation from BackendTrafficPolicy and instead enhanced EnvoyExtensionPolicy by adding a stage field to the ExtProc configuration. This new field allows setting the execution phase to Route (the default), Backend, or All.
This aligns perfectly with your points on security and API consistency and provides a solid foundation for the full implementation.

Let me know what you think of the new direction!

zhaohuabing commented on api/v1alpha1/ext_proc_types.go:
```diff
@@ -24,6 +24,18 @@ const (
 )
 
+// +kubebuilder:validation:Enum=Route;Backend;All
+type ExtProcStage string
```
Should we use a more common name like `ExecuteStage`? It could be used for other filters as well.
